src/nullablearray.jl

```diff
@@ -81,77 +49,27 @@ function NullableCategoricalArray{T, N}(A::AbstractArray{T, N},
-                              ordered::Bool=false)
-
-Similar to definition above, but marks as null entries for which the corresponding entry
-in `missing` is `true`.
-"""
-NullableCategoricalVector{T}(A::AbstractVector{T},
-                             missing::AbstractVector{Bool};
-                             ordered=false) =
-    NullableCategoricalArray(A, missing; ordered=ordered)
-
-"""
-    NullableCategoricalMatrix(A::AbstractMatrix, missing::AbstractMatrix{Bool};
-                              ordered::Bool=false)
-
-Similar to definition above, but marks as null entries for which the corresponding entry
-in `missing` is `true`.
-"""
-NullableCategoricalMatrix{T}(A::AbstractMatrix{T},
-                             missing::AbstractMatrix{Bool};
-                             ordered=false) =
-    NullableCategoricalArray(A, missing; ordered=ordered)
-
-@inline function getindex(A::NullableCategoricalArray, I...)
+@inline function getindex(A::CategoricalArray{T}, I...) where {T>:Null}
     @boundscheck checkbounds(A, I...)
     # Let Array indexing code handle everything
     @inbounds r = A.refs[I...]
 
     if isa(r, Array)
-        return ordered!(arraytype(A)(r, deepcopy(A.pool)),
-                        isordered(A))
+        ret = CategoricalArray{T, ndims(r), eltype(r)}(r, deepcopy(A.pool))
+        return ordered!(ret, isordered(A))
     else
-        S = eltype(eltype(A))
         if r > 0
-            @inbounds return Nullable{S}(A.pool[r])
+            @inbounds return A.pool[r]
         else
-            return Nullable{S}()
+            return null
         end
     end
 end
 
-@inline function setindex!(A::NullableCategoricalArray, v::Nullable, I::Real...)
+@inline function setindex!(A::CategoricalArray{>:Null}, v::Null, I::Real...)
     @boundscheck checkbounds(A, I...)
-    if isnull(v)
-        @inbounds A.refs[I...] = 0
-    else
-        @inbounds A[I...] = get(v)
-    end
-end
-
-levels!(A::NullableCategoricalArray, newlevels::Vector; nullok=false) = _levels!(A, newlevels, nullok=nullok)
-
-droplevels!(A::NullableCategoricalArray) = levels!(A, _unique(Array, A.refs, A.pool))
-
-unique{T}(A::NullableCategoricalArray{T}) = _unique(NullableArray{T}, A.refs, A.pool)
-
-function in{T, N, R}(x::Nullable, y::NullableCategoricalArray{T, N, R})
-    ref = get(y.pool, get(x), zero(R))
-    ref != 0 ? ref in y.refs : false
-end
-
-function in{S<:CategoricalValue, T, N, R}(x::Nullable{S}, y::NullableCategoricalArray{T, N, R})
-    v = get(x)
-    if v.pool === y.pool
-        return v.level in y.refs
-    else
-        ref = get(y.pool, index(v.pool)[v.level], zero(R))
-        return ref != 0 ? ref in y.refs : false
-    end
+    @inbounds A.refs[I...] = 0
 end
 
-in{T, N, R}(x::Any, y::NullableCategoricalArray{T, N, R}) = false
-in{T, N, R}(x::CategoricalValue, y::NullableCategoricalArray{T, N, R}) = false
+in(x::Null, y::CategoricalArray) = false
+in(x::Null, y::CategoricalArray{<:Null}) = !all(v -> v > 0, y.refs)
```

Comment:
I've just noticed I made a mistake here, should be `>:Null` AFAICT. Also needs a test.